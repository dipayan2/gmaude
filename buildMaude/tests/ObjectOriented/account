#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/account.maude -no-banner -no-advise \
  > account.out 2>&1

diff $srcdir/account.expected account.out > /dev/null 2>&1
