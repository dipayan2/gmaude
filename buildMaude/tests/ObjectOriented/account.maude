set show timing off .
set show advisories off .
set verbose on .

  ---- adapted
  ---- from the book All About Maude
  ---- from the Maude Manual
  ---- by the Maude team

omod ACCOUNT is
  protecting QID .
  protecting INT .

  subsort Qid < Oid .
  class Account | bal : Int .
  msgs credit debit : Oid Int -> Msg .
  msg from_to_transfer_ : Oid Oid Int -> Msg .

  vars A B : Oid .
  var  M : Nat .
  vars N N' : Int .

  rl [credit] :
    credit(A, M)
    < A : Account | bal : N >
    => < A : Account | bal : N + M > .

  crl [debit] :
    debit(A, M)
    < A : Account | bal : N >
    => < A : Account | bal : N - M >
    if N >= M .

  crl [transfer] :
    (from A to B transfer M)
    < A : Account | bal : N >
    < B : Account | bal : N' >
    => < A : Account | bal : N - M >
        < B : Account | bal : N' + M >
    if N >= M .
endom

omod SAVING-ACCOUNT is
  including ACCOUNT .
  protecting FLOAT .
  class SavingAccount | rate : Float .
  subclass SavingAccount < Account .
endom

rew < 'A-06238 : Account | bal : 2000 > 
         debit('A-06238, 1000) .

rew < 'A-73728 : SavingAccount | bal : 5000, rate : 3.0 >
    < 'A-06238 : Account | bal : 2000 >
    < 'A-28381 : SavingAccount | bal : 9000, rate : 3.0 >
    debit('A-06238, 1000)
    credit('A-73728, 1300)
    credit('A-28381, 200) .

search in SAVING-ACCOUNT :
            < 'A-73728 : SavingAccount | bal : 5000, rate : 3.0 >
            < 'A-06238 : Account | bal : 2000 >
            < 'A-28381 : SavingAccount | bal : 9000, rate : 3.0 >
            debit('A-06238, 1000)
            credit('A-73728, 1300)
            credit('A-28381, 200)
          =>! C:Configuration 
              < O:Oid : Account | bal : N:Nat > 
          such that N:Nat < 8000 .
