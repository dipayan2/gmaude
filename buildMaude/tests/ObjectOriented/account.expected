Considering object completion on:
  rl [credit] : credit(A, M) < A : Account | bal : N > => < A : Account | bal :
    (N + M) > .
Transformed rule:
  rl [credit] : credit(A, M) < A : V:Account | bal : N, Atts:AttributeSet > =>
    < A : V:Account | bal : (N + M), Atts:AttributeSet > .

Considering object completion on:
  crl [debit] : debit(A, M) < A : Account | bal : N > => < A : Account | bal :
    (N - M) > if N >= M = true .
Transformed rule:
  crl [debit] : debit(A, M) < A : V:Account | bal : N, Atts:AttributeSet > => <
    A : V:Account | bal : (N - M), Atts:AttributeSet > if N >= M = true .

Considering object completion on:
  crl [transfer] : (from A to B transfer M) < A : Account | bal : N > < B :
    Account | bal : N' > => < A : Account | bal : (N - M) > < B : Account | bal
    : (N' + M) > if N >= M = true .
Transformed rule:
  crl [transfer] : (from A to B transfer M) < A : V:Account | bal : N,
    Atts:AttributeSet > < B : V2:Account | bal : N', Atts2:AttributeSet > => <
    A : V:Account | bal : (N - M), Atts:AttributeSet > < B : V2:Account | bal :
    (N' + M), Atts2:AttributeSet > if N >= M = true .

op _`,_ left-identity collapse from AttributeSet to Attribute is unequal.
op __ left-identity collapse from Configuration to Object is unequal.
==========================================
rewrite in SAVING-ACCOUNT : debit('A-06238, 1000) < 'A-06238 : Account | bal :
    2000 > .
rewrites: 3
result Object: < 'A-06238 : Account | bal : 1000 >
==========================================
rewrite in SAVING-ACCOUNT : ((((credit('A-28381, 200) credit('A-73728, 1300))
    debit('A-06238, 1000)) < 'A-28381 : SavingAccount | bal : 9000, rate : 3.0
    >) < 'A-06238 : Account | bal : 2000 >) < 'A-73728 : SavingAccount | bal :
    5000, rate : 3.0 > .
rewrites: 7
result Configuration: < 'A-06238 : Account | bal : 1000 > < 'A-28381 :
    SavingAccount | bal : 9200, rate : 3.0 > < 'A-73728 : SavingAccount | bal :
    6300, rate : 3.0 >
==========================================
search in SAVING-ACCOUNT : ((((credit('A-28381, 200) credit('A-73728, 1300))
    debit('A-06238, 1000)) < 'A-28381 : SavingAccount | bal : 9000, rate : 3.0
    >) < 'A-06238 : Account | bal : 2000 >) < 'A-73728 : SavingAccount | bal :
    5000, rate : 3.0 > =>! C:Configuration < O:Oid : Account | bal : N:Nat >
    such that N:Nat < 8000 = true .

Solution 1 (state 7)
states: 8  rewrites: 29
C:Configuration --> < 'A-28381 : SavingAccount | bal : 9200, rate : 3.0 > <
    'A-73728 : SavingAccount | bal : 6300, rate : 3.0 >
O:Oid --> 'A-06238
N:Nat --> 1000

No more solutions.
states: 8  rewrites: 29
Bye.
